src/Build/Instance/TaskRegistry.cs

```diff
@@ -1149,7 +1149,12 @@ internal class RegisteredTaskRecord : ITranslatable
             /// When ever a taskName is checked against the factory we cache the result so we do not have to
             /// make possibly expensive calls over and over again.
             /// </summary>
-            private Dictionary<RegisteredTaskIdentity, object> _taskNamesCreatableByFactory;
+            private ConcurrentDictionary<RegisteredTaskIdentity, object> _taskNamesCreatableByFactory;
+
+            /// <summary>
+            /// Lock object for initializing the cache
+            /// </summary>
+            private readonly object _taskNamesCreatableByFactoryLock = new object();
```

Comment:
Another option here would be to create the ConcurrentDictionary at construction time--that saves a pointer to an `object` for the lock.

Do we always create a `_taskNamesCreatableByFactory` eventually for every TaskRegistry, or is it worth saving/deferring the dictionary init cost as we're doing today?

Reply:
I now checked where `CanTaskBeCreatedByFactory` is called (where `_taskNamesCreatableByFactory` is getting populated), and it is called in most of the other methods of `RegisteredTaskRecord`. So, yeah it is better to just initialize ConcurrentDictionary at construction time. 
I'll change this. Thanks!